packages/langium/src/lsp/lsp-services.ts

```diff
@@ -0,0 +1,125 @@
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import type { ExecuteCommandHandler } from './execute-command-handler.js';
+import type { CodeActionProvider } from './code-action.js';
+import type { CompletionProvider } from './completion/completion-provider.js';
+import type { DocumentHighlightProvider } from './document-highlight-provider.js';
+import type { DocumentSymbolProvider } from './document-symbol-provider.js';
+import type { FoldingRangeProvider } from './folding-range-provider.js';
+import type { Formatter } from './formatter.js';
+import type { DefinitionProvider } from './definition-provider.js';
+import type { HoverProvider } from './hover-provider.js';
+import type { ReferencesProvider } from './references-provider.js';
+import type { RenameProvider } from './rename-provider.js';
+import type { SemanticTokenProvider } from './semantic-token-provider.js';
+import type { LanguageServer } from './language-server.js';
+import type { SignatureHelpProvider } from './signature-help-provider.js';
+import type { TypeDefinitionProvider } from './type-provider.js';
+import type { ImplementationProvider } from './implementation-provider.js';
+import type { CallHierarchyProvider } from './call-hierarchy-provider.js';
+import type { DocumentLinkProvider } from './document-link-provider.js';
+import type { CodeLensProvider } from './code-lens-provider.js';
+import type { DeclarationProvider } from './declaration-provider.js';
+import type { InlayHintProvider } from './inlay-hint-provider.js';
+import type { WorkspaceSymbolProvider } from './workspace-symbol-provider.js';
+import type { NodeKindProvider } from './node-kind-provider.js';
+import type { FuzzyMatcher } from './fuzzy-matcher.js';
+import type { Connection, TextDocuments } from 'vscode-languageserver';
+import type { TextDocument } from 'vscode-languageserver-textdocument';
+import type { DeepPartial, LangiumCoreServices, LangiumSharedCoreServices } from '../services.js';
+import type { TypeHierarchyProvider } from './type-hierarchy-provider.js';
+import type { DocumentUpdateHandler } from './document-update-handler.js';
+import type { FileOperationHandler } from './file-operation-handler.js';
+import type { ConfigurationProvider } from './../workspace/configuration.js';
+import type { LangiumDocumentFactory } from './../workspace/documents.js';
+import type { WorkspaceManager } from './../workspace/workspace-manager.js';
+import type { ServiceRegistry } from '../service-registry.js';
+
+/**
+ * Combined Core + LSP services of Langium (total services)
+ */
+export type LangiumServices = LangiumCoreServices & LangiumLSPServices;
+
+/**
+ * Combined Core + LSP shared services of Langium (total services)
+ */
+export type LangiumSharedServices = LangiumSharedCoreServices & LangiumSharedLSPServices;
+
+export type LSPServices = {
+    lsp: {
```

Comment:
A followup thought on my previous review: Can we turn `lsp` to `lsp?`? All it's members are optional and need to be checked for being defined, anyway. That would simplify things further.

Reply:
Not if we had it like this before, but I guess it would be fine. Do we get anything by doing this, or is this just going to make us extra checks to verify `lsp` is present?

Reply:
meanwhile invalid.